### ISSUE TITLE
[3.0.0-rc1] Circular logic doesn't result in informative error

### ISSUE THREAD

**Issue by fzembow** [ZJONSSON/clues#8]
**STEPS**
Run the following logic:
```
var logic = {
  A: function(B) {
    return B;
  },

  B: function(A) {
    return A;
  }
};

clues(logic, 'A');
```

**RESULT**
It throws an error, but it's uninformative:
```
/Users/fzembow/git/clues/clues.js:63
    .map(function(arg) {
                 ^
RangeError: Maximum call stack size exceeded
```

**EXPECTED**
An error like "Circular dependency detected" or ideally "Circular dependency detected between A, B"

I ran into this when building something and it was rather annoying to trace down exactly what was happening. Let's talk through how we can address this.

**Comment by ZJONSSON** [comment]
This is still an outstanding issue without an implemented solution.   One way to tackle this is assigning a unique id to each promise (of a function evaluation) and keeping a running tab of "children" for each branch of a "query" (recursive prototype chains for each argument) and compare new IDs to the children each time.  This might hurt performance.  It might make sense to have a "debug mode"  (similar setup as the long stack traces option in bluebird)

**Comment by ZJONSSON** [comment]
A preliminary solution is available under `clues@4.0.0-rc2` - will be merged to master soon.
https://github.com/ZJONSSON/clues/tree/recursive-error#circular-dependencies
https://github.com/ZJONSSON/clues/blob/recursive-error/test/recursion-error-test.js

**Comment by ZJONSSON** [comment]
With `logic = {a:function(b) {}, b: function(a) {}}` the above solution works if you either solve for `a` or solve for `b`.  If you simultaneously solve for both (i.e. `clues(logic,function(a,b) {...})`) the circular nature will not be picked up as `b` is already a promise on a different path/chain when the `a` function reaches for it (and vice-versa).  As we can never guarantee a single path of resolution (as paths fork in a tree) the branch above is pretty useless.

Below is a more extensive solution that works for any number of live paths concurrently - but it is a little inefficient (crawling up the cancellation chain every time you hit an unresolved promise).  Depending on use case, the benefits might outweigh the inefficiencies.   This method is applied when cancellation is turned on in Bluebird, i.e. `Promise.config({cancellation: true})` and `clues.ignoreCircular` has not been explicitly set to `true`
https://github.com/ZJONSSON/clues/commit/5b36c1fcf35b4d8fad51dd0f358089fffcaf8054  (`npm = clues@3.5.0-rc-3`)
